### DISCUSSION TITLE
Regex that accepts any character as unquoted string

### DISCUSSION THREAD

**Discussion by idiosyncrasy00** [eclipse-langium/langium#1010] [General]
Hello, I am trying to build a DSL and there is a piece of work like this:

_sale.isComplete = false
Sequence{prdDesc.desc, prdDesc.price, prdDesc.upc}
_item.salesLineItem->isEmpty()

I want Langium to treat the above expression as an unquoted String. What I have tried so far is to implement a regex that matches any character but langium does not accept me to do this.

How can I solve the problem? Thanks in advance.

**Comment by msujew** [comment]
Hey @LockedAwya,

I'm not sure what you're trying to achieve is easily doable with an LL parser (and therefore with Langium). The issue is that LL parsers first need to tokenize the whole text, before attempting to parse it. During this tokenization step, it cannot take context into account. So it can't differentiate between a normal ID (`Sequence`) and an unquoted String without context.

At least that's what I'm gathering from your question. Maybe you can elaborate a bit on what you're trying to achieve? I can only give you minimal advice with this little context.

**Reply by idiosyncrasy00** [reply]
Basically I got three expressions above from a specific context and what I am trying to do is to treat them as an unquoted string. 
The grammar definition for the terminal of unquoted String is:
`terminal UNQUOTED_STRING:
	'£$%^£$%^'; 
` 
(Not sure if this is a correct regex since I took the reference from a respective Xtext project, and it works just fine)

So far I can only express them as a quoted string like this (This works):
**"_sale.isComplete = false"
"Sequence{prdDesc.desc, prdDesc.price, prdDesc.upc}"
"_item.salesLineItem->isEmpty()"**

What I am trying to achieve is to express them as an unquoted String (This does not work, thus finding a solution):
_sale.isComplete = false 
Sequence{prdDesc.desc, prdDesc.price, prdDesc.upc} 
_item.salesLineItem->isEmpty()

(FYI, the expressions are for later used of an OCL project, thus they should be formatted as an unquoted string).

The problem seems to that Langium only offers an unquoted string as combination of **alphabetical and numerical** characters, not **ANY** characters.

I know it's weird to ask such a question, but I wonder if there is a way that Langium can resolve this.

**Reply by msujew** [reply]
> The problem seems to that Langium only offers an unquoted string as combination of alphabetical and numerical characters, not ANY characters.

Langium is just using the runtime's regex engine to perform tokenization, so you're not limited in any way in that regard. Note that you can always just use the Xtext terminal syntax to express your terminal: Langium will convert it into the corresponding regex syntax.

If you want to use the regex terminal syntax, you'll need to use `//` as delimiters by the way, something like `UNQUOTED_STRING: /£$%^£$%^/;`. However, this does seem a bit weird as a regex, since it only matches the string `£$%^£$%^`.
